[PTD]  Add disable_comm_split_share option to new_group() API

MogicianWu opened the pull request:
Summary:
In genAI training, we want to save NCCL GPU memory by performing comm_split on non-overlapping sub process groups for different parallelisms. However the current API only allows us to create sub process groups by either performing comm split on all of them or on none of them.

Since the creation of new subgroups is executed through new_group() API, adding an additional disable_comm_split flag to allow some subgroups to be created by NcclCommSplit but not others.

Test Plan:
Expanded test_comm_split_subgroup UT test case to cover this use case

```
buck2 test fbcode//mode/opt -c param.use_nccl=True -c fbcode.nvcc_arch=a100 -c hpc_comms.use_ncclx=2.21.5 -c fbcode.enable_gpu_sections=true  fbcode//caffe2/test/distributed:c10d -- -r test_comm_split_subgroup
```

Reviewed By: csmodlin

Differential Revision: D57129913




cc @XilunWu @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o @mrshenli @pritamdamania87 @zhaojuanmao @satgera @gqchen @aazzolini @osalpekar @jiayisuse @penguinwu @tianyu-l @yf225 @chauhang

kwen2501 commented:
> In genAI training, we want to save NCCL GPU memory by performing comm_split on non-overlapping sub process groups for different parallelisms. However the current API only allows us to create sub process groups by either performing comm split on all of them or on none of them.

I failed to catch the "however" relationship between the first sentence and second sentence. Can you please clarify? 

If not too much work, an example would be appreciated -- like, what doesn't work before this PR and what would work after this PR.

MogicianWu commented:
> > In genAI training, we want to save NCCL GPU memory by performing comm_split on non-overlapping sub process groups for different parallelisms. However the current API only allows us to create sub process groups by either performing comm split on all of them or on none of them.
> 
> I failed to catch the "however" relationship between the first sentence and second sentence. Can you please clarify?
> 
> If not too much work, an example would be appreciated -- like, what doesn't work before this PR and what would work after this PR.

My original assumption wasn't correct. We do want to perform NcclCommSplit() on all subgroups, it's just for some subgroups we don't want to share memory resources (which is not offered by current PTD API).

MogicianWu committed:
[PTD] Add disable_comm_split_share option to new_group() API (#125890)

Summary:

In genAI training, we want to save NCCL GPU memory by performing comm_split and turn on split_share option on non-overlapping sub process groups for different parallelisms to save GPU memory. Add this option by exposing an additional 
disable_comm_split_share() option in pg_option so for this particular subgroup NcclCommSplit will have split_share set to False while calling commSplit (default is True)

Test Plan:
Added this API call in test_comm_split_subgroup UT test case; see following diff for integration testing

```
buck2 test fbcode//mode/opt -c param.use_nccl=True -c fbcode.nvcc_arch=a100 -c hpc_comms.use_ncclx=2.21.5 -c fbcode.enable_gpu_sections=true  fbcode//caffe2/test/distributed:c10d -- -r test_comm_split_subgroup
```

Crashed UT and we can see in ProcessGroupNCCL logs P1360775469 splitShare was set to 0 (default value is -2147483648)

Differential Revision: D57129913
